Rust sendrecv

maxmcd opened the pull request:
There's more work to be done here (No flag support, no cleanup, a few shortcuts) but the rust script seems fully functional. Wanted to inquire about two things:

1. Is this rust code welcome (I'm just writing this up as a personal exercise) and is there any feedback?
2. Are the docker-compose and related Dockerfiles desired? I can move them off the branch, but figured they might be helpful. 

Can complete the rust code and clean up/document docker-compose upon confirmation of #2.

maxmcd committed:
Add docker-compose demo

maxmcd committed:
Beginnings of gst-rust, plugin validation and basic WS event loop

maxmcd committed:
Promise and event progress

maxmcd committed:
Add send ice, create offer, send sdp and others

maxmcd committed:
Json message parsing and Sdp message parse_buffer fix

maxmcd committed:
Video to browser working

maxmcd committed:
Add appstate handling with Atomic and Arc

maxmcd committed:
Implement on_incoming_stream, on_incoming_decodebin_stream, handle_media_stream

maxmcd committed:
Remove debugging statements and some cleanup

sdroege commented on sendrecv/gst-rust/src/main.rs:
```diff
@@ -0,0 +1,426 @@
+}
+
+fn sdp_message_as_text(offer: gstreamer_webrtc::WebRTCSessionDescription) -> Option<String> {
+    unsafe {
```
https://github.com/sdroege/gstreamer-rs/issues/108#issuecomment-393505826 for why this is necessary currently. We should fix that before merging this PR :)

sdroege replied:
I've added those now here https://github.com/sdroege/gstreamer-rs/commit/85ca3b94a34d0236314aac900d5926c81ba73d99

maxmcd replied:
Lovely timing, as mine was incorrect: https://github.com/maxmcd/gstreamer-rs/commit/dd0918b05d9cd70c0aa359762080545e1cb1000e

:)

sdroege replied:
A mutable borrow is not needed here as all these are immutable fields after all, and `from_glib_none` is needed here because we don't take ownership of the field (otherwise we would steal it from the struct). Ideally we would just borrow it, but there's no borrowing mechanism in `glib-rs` at this point that could be used for this (and generally borrowing from GLib types is dangerous by default except for some special cases), so we have to copy here.

sdroege replied:
Let me know if my version works and if there are any other problems :) And thanks again for working on this

maxmcd replied:
Your version works perfectly. Replaced the unsafe call.

sdroege commented on sendrecv/gst-rust/src/main.rs:
```diff
@@ -0,0 +1,426 @@
+    return AppState::PeerConnecting;
+}
+
+fn register_with_server(app_state: &Arc<AtomicUsize>, out: &ws::Sender) -> AppState {
```
Might be nicer to move the app state into a struct for extensibility later instead of directly passing around that single integer everywhere (and maybe wrap that struct into an `Rc<Mutex<_>>`)

sdroege commented on sendrecv/gst-rust/src/main.rs:
```diff
@@ -0,0 +1,426 @@
+    webrtc
+        .emit(
+            "set-local-description",
+            &[&offer.to_value(), &promise.to_value()],
```
Hmm what's the point of the promise if it's not used anywhere afterwards?

maxmcd replied:
I was mostly parroting the original c implementation: [webrtc-sendrecv.c#L252-L255](https://github.com/centricular/gstwebrtc-demos/blob/61f49793ed099d165694b3cfb8e8943a7b77723f/sendrecv/gst/webrtc-sendrecv.c#L252-L255) 

It does seem though that no promise is allowed: https://gstreamer.freedesktop.org/data/doc/gstreamer/head/gst-plugins-bad/html/gst-plugins-bad-plugins-webrtcbin.html#GstWebRTCBin-set-local-description

On that point, I couldn't figure out how to pass null in a situation like this. Is there some stand-in in the library?

sdroege replied:
You should be able to pass `&None::<gst::Promise>` here

sdroege commented on sendrecv/gst-rust/src/main.rs:
```diff
@@ -0,0 +1,426 @@
+    });
+    let options = gst::Structure::new_empty("options");
+    webrtc_clone
+        .emit("create-offer", &[&options.to_value(), &promise.to_value()])
```
The `to_value()` in all the `emit` calls should not be needed, only the `&`

sdroege commented on sendrecv/gst-rust/src/main.rs:
```diff
@@ -0,0 +1,426 @@
+        sink.sync_state_with_parent().unwrap();
+        gst::Element::link_many(&[&q, &conv, &resample, &sink]).unwrap();
+    } else {
+        pipe.clone()
```
You could move this `dynamic_cast` outside the `if` and have it only once then, storing the `bin` in a local variable

sdroege commented on sendrecv/gst-rust/src/main.rs:
```diff
@@ -0,0 +1,426 @@
+        queue !
+        application/x-rtp,media=audio,encoding-name=OPUS,payload=97 ! sendrecv.
+        ",
+    ).unwrap();
```
Explicitly creating the elements would seem nicer than string programming, but I guess it's fine :)

sdroege commented on sendrecv/gst-rust/src/main.rs:
```diff
@@ -0,0 +1,426 @@
+        .unwrap();
+
+    let ret = pipe.set_state(gst::State::Playing);
+    assert_ne!(ret, gst::StateChangeReturn::Failure);
```
You could also do `into_result().unwrap()` on the return value

sdroege commented on sendrecv/gst-rust/src/main.rs:
```diff
@@ -0,0 +1,426 @@
+                x if x == AppState::ServerRegistering as usize => AppState::ServerRegisteringError,
+                x if x == AppState::PeerConnecting as usize => AppState::PeerConnectionError,
+                x if x == AppState::PeerConnected as usize => AppState::PeerCallError,
+                x if x == AppState::PeerCallNegotiating as usize => AppState::PeerCallError,
```
See above, if you make the app struct a `Rc<Mutex<_>>` then you could store the state directly as an enum and match properly on it, exhaustively

maxmcd replied:
Can we get away with `Rc` with the promises? I assumed that would behave the same way as spawning a new thread and a `sync::Arc` would be needed.

sdroege replied:
Yes, `Arc` is needed, sorry :)

sdroege commented on sendrecv/gst-rust/src/main.rs:
```diff
@@ -0,0 +1,426 @@
+                .unwrap();
+            self.update_state(AppState::PeerCallStarted);
+        }
+        if json_msg.get("ice").is_some() {
```
Maybe we could use `serde-json` here for directly converting between a Rust struct and JSON?

sdroege commented:
Looks quite nice, thanks :) I have no opinions about the Docker stuff and how the webrtc library is used is for someone else to review.

From the Rust side, in addition to the comments above it might be nice to do proper error handling in various places instead of just unwrapping/expecting

maxmcd commented:
Thanks so much for the detailed feedback @sdroege!

maxmcd committed:
Added minor fixes and improvements related to PR

maxmcd committed:
Add some error handling, add more to Mutex, minor fixes

maxmcd committed:
Add peer-id and server flags, clean up some mutex use, general cleanup

maxmcd committed:
Remove unsafe fetching of sdp as text

maxmcd commented:
I'm done with edits for the moment. The mutex stuff is a little messy in places and it's still quite panicky. Had a difficult time consolidating the async nature of the code with proper error handling. 

Also, cleaned up docker and added peer-id/server flag support to gain parity with the c version.

sdroege commented:
So this is working and basically equivalent with the C version now, and could get another full review?

nirbheek commented on signalling/Dockerfile:
```diff
@@ -0,0 +1,13 @@
+RUN sed -i 's/sslctx.load_cert_chain(chain_pem, keyfile=key_pem)/pass/g' \
+    ./simple-server.py
+RUN sed -i 's/ssl=sslctx,//g' \
+    ./simple-server.py
```
Maybe you want to add an option to `simple-server.py` to not use ssl instead? If you set `ssl=None` when calling `serve()`, it's the same as not passing an sslctx.

nirbheek commented on sendrecv/gst/Dockerfile:
```diff
@@ -0,0 +1,20 @@
+RUN sed -i 's/wss:\/\/webrtc.nirbheek.in:8443/ws:\/\/signalling:8443/g' \
+    /opt/webrtc-sendrecv.c
+RUN sed -i 's/strict_ssl = TRUE/strict_ssl = FALSE/g' \
+    /opt/webrtc-sendrecv.c
```
Here, too. I think it would be generally useful for people (while running local servers) if we do that.

maxmcd commented:
@sdroege yes absolutely

sdroege commented:
Great, I'll do so later then :)

maxmcd committed:
Add --disable-ssl option to simple-server.py

maxmcd committed:
Add --disable-ssl flag to webrtc-sendrecv.c

sdroege commented on sendrecv/gst-rust/src/main.rs:
```diff
@@ -0,0 +1,480 @@
+extern crate serde_json;
+
+use gst::prelude::*;
+use gst::{BinExt, ElementExt};
```
This line should be unneeded, `gst::prelude::*` already imports all the ext traits into the scope

sdroege commented on sendrecv/gst-rust/src/main.rs:
```diff
@@ -0,0 +1,480 @@
+extern crate glib;
+extern crate gstreamer as gst;
+extern crate gstreamer_sdp;
+extern crate gstreamer_sdp_sys;
```
This should be unneeded now

sdroege commented on sendrecv/gst-rust/src/main.rs:
```diff
@@ -0,0 +1,480 @@
+extern crate clap;
+extern crate glib;
+extern crate gstreamer as gst;
+extern crate gstreamer_sdp;
```
I'd recommend importing this as `gst_sdp` (and below `gst_webrtc`) for consistency and less typing

sdroege commented on sendrecv/gst-rust/src/main.rs:
```diff
@@ -0,0 +1,480 @@
+enum Error {
+    GlibError(glib::Error),
+    GlibBoolError(glib::BoolError),
+    GstStateChangeError(gst::StateChangeError),
```
Instead of this you could also use the `failure` crate and just work with `failure::Error` everywhere (which every error can be converted to, e.g. via the ?-operator). Unless you want to specifically match and handle specific errors, this would simplify the code. See the `gstreamer-rs` examples for some example usage

sdroege commented on sendrecv/gst-rust/src/main.rs:
```diff
@@ -0,0 +1,480 @@
+
+const STUN_SERVER: &'static str = "stun-server=stun://stun.l.google.com:19302 ";
+const RTP_CAPS_OPUS: &'static str = "application/x-rtp,media=audio,encoding-name=OPUS,payload=";
+const RTP_CAPS_VP8: &'static str = "application/x-rtp,media=video,encoding-name=VP8,payload=";
```
Personally I prefer creating caps explicitly instead of from a string. You could e.g. have functions here that take the remaining missing parameters and then return a `gst::Caps` via `gst::Caps::new_simple()`

sdroege commented on sendrecv/gst-rust/src/main.rs:
```diff
@@ -0,0 +1,480 @@
+const RTP_CAPS_OPUS: &'static str = "application/x-rtp,media=audio,encoding-name=OPUS,payload=";
+const RTP_CAPS_VP8: &'static str = "application/x-rtp,media=video,encoding-name=VP8,payload=";
+
+fn check_plugins() -> bool {
```
This probably should return a `Result<(), SomeErrorType>`, and that error type could contain the element name. See e.g. https://github.com/sdroege/gstreamer-rs/blob/85ca3b94a34d0236314aac900d5926c81ba73d99/examples/src/bin/appsrc.rs#L19-L21

sdroege commented on sendrecv/gst-rust/src/main.rs:
```diff
@@ -0,0 +1,480 @@
+        .ws_sender
+        .send(format!("SESSION {}", app_control.peer_id))
+        .unwrap();
+    return AppState::PeerConnecting;
```
Don't use the `return` keyword in such cases. clippy should also warn about this

sdroege commented on sendrecv/gst-rust/src/main.rs:
```diff
@@ -0,0 +1,480 @@
+
+fn setup_call(app_control: &Arc<Mutex<AppControl>>) -> AppState {
+    let mut app_control = app_control.lock().unwrap();
+    app_control.app_state = AppState::PeerConnecting;
```
Why is the (same) state set and returned here (and in the other such functions)?

sdroege commented on sendrecv/gst-rust/src/main.rs:
```diff
@@ -0,0 +1,480 @@
+    app_control
+        .ws_sender
+        .send(format!("HELLO {}", our_id))
+        .unwrap();
```
Is the websocket API synchronous? Might make sense to do this asynchronously to not block any further processing

maxmcd replied:
Looks like it's synchronous. I'm unfamiliar with how to simply make a call async in rust without throwing it in a thread. Is there something in glib that could be leveraged? Or it seems the futures crate is used in gst-rs.

sdroege replied:
Don't worry about that for now and just add a FIXME comment to that effect.

You can use glib for async things if the websocket implementation is integrated with the glib async infrastructure (which it isn't). The `ws` crate looks completely synchronous, I think what you need to do actually is to run `ws::connect()` on another thread and then communicate with that, and have another thread that does the GStreamer bus handling (and maybe generally uses the glib main loop).

Note this part on the `ws::connect()` function: `This function blocks until the event loop finishes running.`

Currently you would block the main thread when calling that function until all is done. There must be a way to somehow communicate from another thread with the socket or the handler so that you can integrate this with other application parts. This needs some changes and work, and probably also solves automatically your question about the bus handling. The answer is that you need to somehow have one event loop for the websocket and another for the bus (and potential other GLib things), or you have to integrate them with each other somehow. You need to check what mechanisms the `ws` crate provides for that.

sdroege commented on sendrecv/gst-rust/src/main.rs:
```diff
@@ -0,0 +1,480 @@
+    let app_control = app_control.lock().unwrap();
+    if app_control.app_state < AppState::PeerCallNegotiating {
+        // TODO signal and cleanup
+        panic!("Can't send offer, not in call");
```
In the context of GStreamer what I usually do for reporting errors from somewhere in other threads to the main event loop is by posting a message on the bus. This could also transfer some Rust error type around.

See e.g. https://github.com/sdroege/gstreamer-rs/blob/85ca3b94a34d0236314aac900d5926c81ba73d99/examples/src/bin/decodebin.rs#L131-L139 and https://github.com/sdroege/gstreamer-rs/blob/85ca3b94a34d0236314aac900d5926c81ba73d99/examples/src/bin/decodebin.rs#L167-L185

sdroege commented on sendrecv/gst-rust/src/main.rs:
```diff
@@ -0,0 +1,480 @@
+        "type": "offer",
+        "sdp": offer.get_sdp().as_text().unwrap(),
+      }
+    });
```
For all these JSON things you could define Rust structs, use `serde-derive` and then let `serde` automatically serialize/deserialize them for you instead of having to manually construct/parse JSON. That should make this code a bit more beautiful

maxmcd replied:
I got a little held up here because `type` is a reserved word :)
But yes, agreed, will add those

sdroege replied:
```rust
struct Bla {
    #[serde(rename = "type")]
    type_: String,
}
```

Should work

maxmcd replied:
I didn't realize initially that serde can cast to an enum, the fix here is super clean

sdroege commented on sendrecv/gst-rust/src/main.rs:
```diff
@@ -0,0 +1,480 @@
+        "sdp": offer.get_sdp().as_text().unwrap(),
+      }
+    });
+    app_control.ws_sender.send(message.to_string()).unwrap();
```
Most of these unwraps could be converted to the ?-operator and then at some point you convert those into a GStreamer error message that is posted on the bus

sdroege commented on sendrecv/gst-rust/src/main.rs:
```diff
@@ -0,0 +1,480 @@
+    let webrtc = values[0].get::<gst::Element>().expect("Invalid argument");
+    let webrtc_clone = webrtc.clone();
+    let app_control_clone = app_control.clone();
+    let promise = gst::Promise::new_with_change_func(move |promise: &gst::Promise| {
```
Why is it required to provide the closure parameter type here?

sdroege commented on sendrecv/gst-rust/src/main.rs:
```diff
@@ -0,0 +1,480 @@
+    });
+    let options = gst::Structure::new_empty("options");
+    webrtc_clone
+        .emit("create-offer", &[&options, &promise])
```
Maybe the options could also be `None` instead of an empty new structure?

sdroege commented on sendrecv/gst-rust/src/main.rs:
```diff
@@ -0,0 +1,480 @@
+    let conv = gst::ElementFactory::make(convert_name, None).unwrap();
+    let sink = gst::ElementFactory::make(sink_name, None).unwrap();
+
+    let pipe_bin = pipe.clone().dynamic_cast::<gst::Bin>().unwrap();
```
Instead of dynamic casting here, you could do that already (to a `gst::Pipeline` I guess?) where you connect to the `pad-added` signal. That way you only need the cast once down there instead of in all these functions everwhere.

sdroege commented on sendrecv/gst-rust/src/main.rs:
```diff
@@ -0,0 +1,480 @@
+        conv.sync_state_with_parent()?;
+        resample.sync_state_with_parent()?;
+        sink.sync_state_with_parent()?;
+        gst::Element::link_many(&[&q, &conv, &resample, &sink])?;
```
These lines should be swapped. Add to the bin, link, sync state

sdroege commented on sendrecv/gst-rust/src/main.rs:
```diff
@@ -0,0 +1,480 @@
+
+    let pipe_bin = pipe.clone().dynamic_cast::<gst::Bin>().unwrap();
+
+    if convert_name == "audioconvert" {
```
Maybe instead of stringly-typed programming, there could be an `enum MediaType { Audio, Video }`?

sdroege commented on sendrecv/gst-rust/src/main.rs:
```diff
@@ -0,0 +1,480 @@
+
+struct WsClient {
+    webrtc: Option<gst::Element>,
+    app_control: Arc<Mutex<AppControl>>,
```
I would do this as follow
```rust
#[derive(Clone)]
struct AppControl(Arc<Mutex<AppControlInner>>);`

struct AppControlInner {
   [the fields from below]
}
```

And then pass a `&AppControl` to all the functions... or maybe even make all the functions you have above part of an `impl AppControl` block

sdroege replied:
Also should there be maybe only one struct instead of a separate `WsClient` that only has the element in addition?

sdroege commented on sendrecv/gst-rust/src/main.rs:
```diff
@@ -0,0 +1,480 @@
+}
+
+struct WsClient {
+    webrtc: Option<gst::Element>,
```
Why `Option`, shouldn't this always exist?

maxmcd replied:
`WsClient` needs to be passed to `ws::connect` and `webrtc` is created in `on_open`, even with the new Mutex structure I think this gap should still exist. It does seem a little silly. Is there a better way to handle?

sdroege replied:
It seems fine. You could create a pipeline from the beginning and only add the other elements later, but just having the `Option` here seems fine too.

In theory you could make your `struct` into an `enum`, first with a variant without pipeline and then a variant with the pipeline, but that seems more work than worth it.

sdroege commented on sendrecv/gst-rust/src/main.rs:
```diff
@@ -0,0 +1,480 @@
+        Ok(())
+    }
+
+    fn on_message(&mut self, msg: ws::Message) -> ws::Result<()> {
```
Is there also some kind of `on_error` or so that would be called if the connection attempt fails or anything else goes wrong?

sdroege commented on sendrecv/gst-rust/src/main.rs:
```diff
@@ -0,0 +1,480 @@
+                panic!("ERROR: Received SESSION_OK when not calling");
+            }
+            self.update_state(AppState::PeerConnected);
+            self.webrtc = match start_pipeline(&self.app_control) {
```
The assertion, `update_state` and `start_pipeline` seem like something that could all be moved into a single function inside a `impl AppControl` block (after refactoring as mentioned above). That way you have one function for each message, the assertions locally in there, and here you only dispatch based on the message type

sdroege commented:
> The mutex stuff is a little messy in places

I believe that would be nicer when refactoring how the structs are looking

> and it's still quite panicky

`unwrap` or `expect` is fine for things that can't possibly fail, for everything else maybe the pattern of converting to a GStreamer error message on the bus could help (also you really should listen to message on the bus, see https://github.com/sdroege/gstreamer-rs/blob/85ca3b94a34d0236314aac900d5926c81ba73d99/examples/src/bin/launch_glib_main.rs#L28-L47 for an example).

To have a GStreamer bus available all the time, it might make sense to create an empty pipeline in `main` (which then has a bus) and only add the other parts of the pipeline to it at the later time.

maxmcd commented:
> To have a GStreamer bus available all the time, it might make sense to create an empty pipeline in main (which then has a bus) and only add the other parts of the pipeline to it at the later time.

@sdroege Not that it's really applicable here, but would that work with multiple connections? ie: is it a shared bus among multiple pipelines or one bus for one new pipeline?

sdroege commented:
> @sdroege Not that it's really applicable here, but would that work with multiple connections? ie: is it a shared bus among multiple pipelines or one bus for one new pipeline?

Each pipeline has its own bus, but you could handle multiple busses on the same GLib main loop just fine. You could also create a new bus yourself independent from any pipelines if you want.

sdroege commented:
Btw, don't hesitate to ask if you need any help with anything or are stuck somewhere :)

maxmcd committed:
Add minor fixes, manually construct initial pipeline

maxmcd commented:
Added some minor fixes. Is `set_property_from_str` acceptable when constructing the pipeline? As an example: seems like I would need `GstAudioTestSrcWave` to be imported to properly replace: `audiotestsrc.set_property_from_str("wave", "red-noise");`

maxmcd committed:
Use structs for json handling

maxmcd committed:
Combine WsClient and AppControl

maxmcd commented:
Having some trouble thinking about error handling and the bus. I can set a bus on the main loop with a main pipeline and add the webrtcbin pipeline as a child. At any point I'm going to need to identify an error, write to the bus and close and clean up the websocket connection. Should I just define a method on `AppControl` that handles writing the bus error and cleaning up the connection? Then quit the program once the error has propagated to the main bus? I could use the bus alone if I didn't worry about cleaning up the connection. 

I guess, any additional clarification on what you thought the chain of events would be upon identifying an error?

sdroege commented:
> Added some minor fixes. Is set_property_from_str acceptable when constructing the pipeline? As an example: seems like I would need GstAudioTestSrcWave to be imported to properly replace: audiotestsrc.set_property_from_str("wave", "red-noise");

That's fine. The alternative would be something like this https://github.com/sdroege/gstreamer-rs/blob/f4d57a66522183d4927b47af422e8f321182111f/examples/src/bin/playbin.rs#L27-L35 (there's basically the same API for enums)

sdroege commented:
For how to integrate `GstBus` and proper error handling here, see https://github.com/centricular/gstwebrtc-demos/pull/20#discussion_r194971932. That's a general architectural problem right now with how the `ws` crate is used.

sdroege commented:
> For how to integrate GstBus and proper error handling here, see #20 (comment). That's a general architectural problem right now with how the ws crate is used.

You could basically have another thread that does the ws stuff, and the main thread runs the `GstBus` main loop. And from the ws handler you could put messages on the bus also, or otherwise call things on the main thread via the glib main context API.

The other way around, calling things from elsewhere into the ws thread/event loop does not seem to be possible from a short look over the API. But maybe you also don't need that?

Otherwise there are some tokio based websocket libraries that might be possible to use, but that complicates everything quite a lot :)

maxmcd committed:
Minor refactoring and cleanup

maxmcd commented:
Not sure how I'm missing this. How would I create a message for the bus and what MessageView type would you recommend using for posting websocket messages to the bus?

sdroege commented:
> Not sure how I'm missing this. How would I create a message for the bus and what MessageView type would you recommend using for posting websocket messages to the bus?

You would use `bus.post()` for posting the message, and you could either create an application message, or directly an error. Application would probably be nicer.

See https://github.com/sdroege/gstreamer-rs/blob/db3fe694154c697afdaf3efb6ec65332546942e0/tutorials/src/bin/basic-tutorial-5.rs#L264-L267 for an example involving an element, you can simply do `bus.post(msg)` in your case instead of using `element.post_message(msg)`. The structure from the application message can contain anything you want apart from its name.

You can catch the message on the bus like https://github.com/sdroege/gstreamer-rs/blob/db3fe694154c697afdaf3efb6ec65332546942e0/tutorials/src/bin/basic-tutorial-5.rs#L241-L244, for example.

maxmcd committed:
Replace ws lib, use bus and better error handling, remove all panics

maxmcd commented:
Rewrote the script with rust-websocket and have better error handling with no panics and (I think) much less blocking. Much happier with where this is now, doesn't seem to be fighting against itself as much.

The error handling within the callback closures is a little hacky, and looking at some of the gstreamer-rs examples my bus/loop setup could be a little cleaner.

Ready for another round of feedback/review.

maxmcd committed:
Cleanup and more idiomatic error handling for missing elements

maxmcd committed:
Use enum for json serialization as well

maxmcd committed:
Better handle signalling server going away

sdroege commented on sendrecv/gst-rust/src/main.rs:
```diff
@@ -0,0 +1,768 @@
+            MessageView::Element(_) => {}
+            MessageView::Tag(_) => {}
+            MessageView::NewClock(_) => {}
+            MessageView::Warning(warning) => {
```
You also want to handle `Error` messages, but all the others above are not really needed :) An `Error` message is always fatal

sdroege commented on sendrecv/gst-rust/src/main.rs:
```diff
@@ -0,0 +1,768 @@
+            println!("Got error message! {}", msg);
+            return self.handle_error();
+        }
+        let json_msg: JsonMsg = serde_json::from_str(&msg).unwrap();
```
Instead of unwrapping, failing to deserialize the JSON should probably just be an error like other errors

sdroege commented:
Looks very good, apart from the minor things above I didn't see anything wrong while shortly looking over it. Will do another proper review in the next days but this looks like it's done. Much better with the other websocket library, and the need for 2 threads instead of 1 for that can probably be refactored later (e.g. by using the tokio-based async API on it). But that's for later

sdroege commented on sendrecv/gst-rust/src/main.rs:
```diff
@@ -0,0 +1,768 @@
+        }
+    }
+
+    fn app_state_lt(&self, state: AppState, error_msg: &'static str) -> bool {
```
Maybe this could mirror the `Ord` trait a bit more and be called `app_state_cmp` and return a `std::cmp::Ordering`. But details. Just `lt` seems a bit short :)

maxmcd committed:
Move caught bus message types to catchall, lt => cmp, handle bus errors

maxmcd committed:
Consolidate app_control cloning

sdroege reviewed (approved):
Looks good to me, thanks a lot for your work :)

sdroege commented:
All merged now :)

maxmcd commented:
Nice! Thanks so much for your patience and feedback @sdroege

MathieuDuponchelle commented:
@maxmcd , nice work, as far as I can tell this only works when the signalling server is launched with `--disable-ssl` , or am I missing something?

maxmcd commented:
@MathieuDuponchelle correct, at least that's what I used when I worked on this, I'm sure you could get ssl working again. This is somewhat documented here: https://github.com/centricular/gstwebrtc-demos/blob/master/signalling/Dockerfile

The priority was getting a quick docker demo working so would check the dockerfiles and docker-compose file for quick reference on how to run.

sdroege commented:
This will need some refactoring to do all the ws communication on a single thread via tokio/futures instead of splitting the socket in its sender and receiver side

sdroege commented:
@maxmcd would you be interested in working on this? In any case, we should probably open a new issue for this :)

maxmcd commented:
Yep, would be happy to.

sdroege commented:
Great, let us know how it goes and if you need help or want someone else to take over
